atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/ReloadingCqlClusterContainer.java

```diff
@@ -0,0 +1,134 @@
+ * Consumers should be resilient to any uses after the cluster has been closed, and where necessary, retry.
+ *
+ * After {@link #close()} has returned, no further CQL Clusters will be created, and all managed CQL Clusters will be
+ * closed.
+ */
+public final class ReloadingCqlClusterContainer implements Closeable, Supplier<CqlCluster> {
+    private static final SafeLogger log = SafeLoggerFactory.get(ReloadingCqlClusterContainer.class);
+
+    private final AtomicReference<Optional<CqlCluster>> lastCqlCluster;
+    private final Refreshable<CqlCluster> refreshableCqlCluster;
+    private boolean isClosed;
+
+    private ReloadingCqlClusterContainer(
+            CassandraClusterConfig cassandraClusterConfig,
+            Refreshable<CassandraServersConfig> refreshableCassandraServersConfig,
+            Namespace namespace,
+            CqlClusterFactory cqlClusterFactory) {
+        this.isClosed = false;
+        this.lastCqlCluster = new AtomicReference<>(Optional.empty());
+        this.refreshableCqlCluster = refreshableCassandraServersConfig.map(cassandraServersConfig ->
+                createNewCluster(cassandraClusterConfig, cassandraServersConfig, namespace, cqlClusterFactory));
+
+        refreshableCqlCluster.subscribe(cqlCluster -> {
+            Optional<CqlCluster> maybeCqlClusterToClose = lastCqlCluster.getAndSet(Optional.of(cqlCluster));
+            if (maybeCqlClusterToClose.isPresent()) {
+                try {
+                    maybeCqlClusterToClose.get().close();
+                } catch (IOException e) {
+                    log.warn("Failed to close CQL Cluster after reloading server list", e);
+                }
+            }
+        });
+    }
+
+    public static ReloadingCqlClusterContainer of(
+            CassandraClusterConfig cassandraClusterConfig,
+            Refreshable<CassandraServersConfig> refreshableCassandraServersConfig,
+            Namespace namespace) {
+        return of(cassandraClusterConfig, refreshableCassandraServersConfig, namespace, CqlCluster::create);
+    }
+
+    public static ReloadingCqlClusterContainer of(
+            CassandraClusterConfig cassandraClusterConfig,
+            Refreshable<CassandraServersConfig> refreshableCassandraServersConfig,
+            Namespace namespace,
+            CqlClusterFactory cqlClusterFactory) {
+        return new ReloadingCqlClusterContainer(
+                cassandraClusterConfig, refreshableCassandraServersConfig, namespace, cqlClusterFactory);
+    }
+
+    /**
+     * Synchronized: See {@link #close()}.
+     */
+    private synchronized CqlCluster createNewCluster(
+            CassandraClusterConfig cassandraClusterConfig,
+            CassandraServersConfig cassandraServersConfig,
+            Namespace namespace,
+            CqlClusterFactory cqlClusterFactory) {
+        if (isClosed) {
+            throw new IllegalStateException();
+        }
```

Comment:
can you make this a `SafeIllegalStateException` with some information in the exception message (and perhaps safe arg anything relevant?)